### ISSUE TITLE
Bidirectional RNNs

### ISSUE THREAD

**Issue by alquraishi** [tensorflow/tensorflow#403]
I'd like to request support for bidirectional RNNs, as these are rather canonical architectures. I realize it's possible to do using the existing rnn API, but given all the intricacies of padding and such (especially in the reverse direction), it would be nice if this had builtin support.

**Comment by ludimagister** [comment]
Good request! Bidirectional RNN code is checked in internally, will go out some time soon with the next release (within a few weeks).

**Comment by LeavesBreathe** [comment]
Is the code within github now? or is it still to come as you said in the next release?

**Comment by vrv** [comment]
Our policy is that until it is documented, assume it is private and not ready to be used.

@ludimagister: let's reopen this until it's actually done and public

**Comment by LeavesBreathe** [comment]
Okay great! That's a pretty smart policy. Thanks @vrv

**Comment by alquraishi** [comment]
Will it have builtin supporting for padding sequences of different length? This is easy to do in the one-directional case but the bidirectional case has to flip the padding and align the activations of the shorter sequences.

**Comment by ludimagister** [comment]
Yes, for the backward RNN it will internally reverse the input sequences up to a specified length (per sequence for the given batch) and reverse then this output to produce the backward output which is then concatenated with the forward output, so the padding will be correct. The unit test should make this clear once the code is pushed.

**Comment by raingo** [comment]
@ludimagister 

Seems like the BiRNN has merged. A question: Does it support padding on the front? For example, if a sequence with length 3 is padded to length 5, the layout is like this:

PAD, PAD, TOKEN1, TOKEN2, TOKEN3

If I understand correctly, the backward RNN is not expecting input like this. Please correct me, if I am wrong.

**Comment by ludimagister** [comment]
No, it doesn't support padding at the front, it only supports to pass in the lengths of each individual sequence in the batch, and will do the correct forward/backward pass on them, up to the specified length. Since every sequence is independent so I am not sure why it would make sense to support aligning them on the right (we picked to align them on the left).

**Comment by raingo** [comment]
@ludimagister Padding on the front (aligning on the right) is usually used for the inputs of the sequence to sequence model.

**Comment by ebrevdo** [comment]
I plan to add support for this but it won't happen until at least January.

**Comment by Smerity** [comment]
Even though this is not yet public / documented I have a small note. Sequence length should not be required for `bidirectional_rnn` as `_reverse_seq` simply reverses the input if `lengths` is not given. This trivially covers cases where the input is already padded or where the input is a static number of time steps.

```
   if not isinstance(inputs, list):
     raise TypeError("inputs must be a list")
-  if not sequence_length:
-    raise ValueError("sequence_length has to be defined")
   if not inputs:
     raise ValueError("inputs must not be empty")
```

**Comment by ebrevdo** [comment]
Yes we should remove this.  +ludigmaster@
On Dec 16, 2015 5:17 PM, "Stephen Merity" <notifications@github.com> wrote:

Even though this is not yet public / documented I have a small note.
Sequence length should not be required for bidirectional_rnn as _reverse_seq
simply reverses the input if lengths is not given. This trivially covers
cases where the input is already padded or where the input is a static
number of time steps.

   if not isinstance(inputs, list):
     raise TypeError("inputs must be a list")
-  if not sequence_length:
-    raise ValueError("sequence_length has to be defined")
   if not inputs:
     raise ValueError("inputs must not be empty")

—
Reply to this email directly or view it on GitHub
<https://github.com/tensorflow/tensorflow/issues/403#issuecomment-165325888>
.

**Comment by ludimagister** [comment]
@Smerity Yes, very good point, will fix tomorrow.

**Comment by shoaibahmed** [comment]
It has been quite a long time since we have been waiting for the Bidirectional RNN. When will it be made available to public?

**Comment by alquraishi** [comment]
@shoaibahmed even though they haven't been officially made public yet, bidirectional RNNs are available in rnn.py, along with the other preliminary RNN architectures.

**Comment by shoaibahmed** [comment]
Yes, I have checked that out @alquraishi . I want to use it for sequence classification purposes i.e. assigning a single label to a whole sequence and the example described there is very different.

**Comment by Mistobaan** [comment]
Any good paper that you would recommend explaining bi-directional RNN ?

**Comment by ludimagister** [comment]
On Tue, Dec 29, 2015 at 10:37 AM, Fabrizio Milo <notifications@github.com>
wrote:

> Any good paper that you would recommend explaining bi-directional RNN ?
>
http://lmgtfy.com/?q=bidirectional+recurrent+neural+networks

**Comment by drdozer** [comment]
The example I'd love to see is how to use BiRNN to compute a sequence with one element per input sequence, with an architecture like this:

```
      /-- F_0..t --\
X_0..t            -- fullyConnected_0..t --> Y_0..t
      \-- R_0..t --/
```

So to predict each element Y_i, an instance of a fully connected network FC_i is attached to F_i and R_i, with all FC sharing all weights.

**Comment by endeavorui** [comment]
I am wondering what is going on with the bidirectional RNN function in TF. I am currently working with bidirectional RNN but asking for different batch_size in training and testing. Following https://github.com/aymericdamien/TensorFlow-Examples/blob/master/notebooks/3%20-%20Neural%20Networks/bidirectional_rnn.ipynb, I am now facing the problem to set seq_len = batch_size. Any suggestions?

**Comment by shoaibahmed** [comment]
RNNs are available in python code. How can we create an RNN using C++ API in TensorFlow?

**Comment by girving** [comment]
@ludimagister, @ebrevdo: What's the status of bidirectional RNNs?

**Comment by s4sarath** [comment]
Hi , A small doubt . In Bi - directional RNN if the input is [ w1 , w2 , w3 , w4 , PAD , PAD ] , while its reversing internally , will it reverse like [ PAD , PAD , w4 , w3 , w2 , w1 ] . The doubt is because , while concatenating the forward hidden state and backward hidden state ( reversed ) , if the hidden state computes in the backward RNN is starting from left to right , we will lose the information . ( hidden state fw(w1) + hidden state bw(PAD) ) . This is not what we expect I guess . I have not seen the implementation , just want to know if it is processing correctly or not .

**Comment by ebrevdo** [comment]
Provide sequence_lengths and it will reverse correctly, keeping padding on
the right.

**Comment by danijar** [comment]
You can also compute them dynamically if your sequences are zero-padded (more details [on my blog][1]):

```python
mask = tf.sign(tf.reduce_max(tf.abs(sequence), 2))
length = tf.reduce_sum(mask, 1)
```

[1]: https://danijar.com/variable-sequence-lengths-in-tensorflow/

**Comment by s4sarath** [comment]
@danijar . Yeah I am following your blog for variable length sequence classification. 
@ebrevdo  - Thanks .

**Comment by ebrevdo** [comment]
@danlijar you should change the following line on your blog:

max_length = int(output.get_shape()[1])

to:

max_length = tf.shape(output)[1]

since in general max_length can vary from run to run.

**Comment by danijar** [comment]
@ebrevdo Thanks for your feedback :)

**Comment by michaelisard** [comment]
@danijar @ebrevdo is this still an open feature request? If the code is now out I will close and move usage discussions to StackOverflow.

**Comment by suisuiwudi** [comment]
Do we get an RNN using C++ API in TensorFlow?
